components/renku_data_services/connected_services/blueprints.py

```diff
@@ -0,0 +1,213 @@
+from renku_data_services.connected_services.apispec_base import AuthorizeParams
+from renku_data_services.connected_services.db import ConnectedServicesRepository
+
+
+@dataclass(kw_only=True)
+class OAuth2ClientsBP(CustomBlueprint):
+    """Handlers for using OAuth2 Clients."""
+
+    connected_services_repo: ConnectedServicesRepository
+    authenticator: base_models.Authenticator
+
+    def get_all(self) -> BlueprintFactoryResponse:
+        """List all OAuth2 Clients."""
+
+        @authenticate(self.authenticator)
+        async def _get_all(_: Request, user: base_models.APIUser):
+            clients = await self.connected_services_repo.get_oauth2_clients(user=user)
+            return json(
+                [apispec.Provider.model_validate(c).model_dump(exclude_none=True, mode="json") for c in clients]
+            )
+
+        return "/oauth2/providers", ["GET"], _get_all
+
+    def get_one(self) -> BlueprintFactoryResponse:
+        """Get a specific OAuth2 Client."""
+
+        @authenticate(self.authenticator)
+        async def _get_one(_: Request, provider_id: str, user: base_models.APIUser):
+            client = await self.connected_services_repo.get_oauth2_client(provider_id=provider_id, user=user)
+            return json(apispec.Provider.model_validate(client).model_dump(exclude_none=True, mode="json"))
+
+        return "/oauth2/providers/<provider_id>", ["GET"], _get_one
+
+    def authorize(self) -> BlueprintFactoryResponse:
+        """Authorize an OAuth2 Client."""
+
+        @authenticate(self.authenticator)
+        @only_authenticated
+        async def _authorize(request: Request, provider_id: str, user: base_models.APIUser):
+            params = AuthorizeParams.model_validate(dict(request.query_args))
+            callback_url = self._get_callback_url(request)
+            url, cookie = await self.connected_services_repo.authorize_client(
+                provider_id=provider_id, user=user, callback_url=callback_url, next_url=params.next
+            )
+            response = redirect(to=url)
+            response.add_cookie("renku-oauth2", cookie, httponly=True)
+            return response
+
+        return "/oauth2/providers/<provider_id>/authorize", ["GET"], _authorize
+
+    def authorize_callback(self) -> BlueprintFactoryResponse:
+        """OAuth2 authorization callback."""
+
+        async def _callback(request: Request):
+            cookie = request.cookies.get("renku-oauth2")
+            cookie = cookie if cookie else ""
+
+            params = AuthorizeParams.model_validate(dict(request.query_args))
+
+            callback_url = self._get_callback_url(request)
+            next_url = params.next
+
+            await self.connected_services_repo.authorize_callback(
+                cookie=cookie, raw_url=request.url, callback_url=callback_url, next_url=next_url
+            )
+
+            response = redirect(to=next_url) if next_url else json({"status": "OK"})
```

Comment:
In addition it would be also nice to test what kind of message we display if the login fails or something is misconfigured. The oauth2 standard has specific rules about how it relays error messages for this case. It would just be nice to show something human readable.

Reply:
A user is going to be redirected to `next_url`, so they would not see a page here. Code maintainer may test the endpoint without a `next_url`, so a simple "OK" message is shown in this case.